src/mscorlib/corefx/System/Runtime/InteropServices/SafeHeapHandleCache.cs

```diff
@@ -0,0 +1,97 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Threading;
+
+namespace System.Runtime.InteropServices
+{
+    /// <summary>
+    /// Allows limited thread safe reuse of heap buffers to limit memory pressure.
+    /// 
+    /// This cache does not ensure that multiple copies of handles are not released back into the cache.
+    /// </summary>
+    internal sealed class SafeHeapHandleCache : IDisposable
```

Comment:
What isn't available in CoreRT?

There were a few things I was taking advantage of that will require some thought/work:

1. The ability to resize in place and retain content. It is quite normal, in fact, for Windows to not even need to move the buffer you're using- avoids copy hits.
2. Lock free caching.
3. Dropping "large" buffers to avoid excessive memory pressure as Path buffers in particular can grow to 64K. (Could just wrap ArrayPool for that.)